### ISSUE TITLE
Balrog accepts scheduled rule changes with duplicate alias'

### ISSUE THREAD

**Issue by bhearsum** [mozilla-releng/balrog#2942]
[Alias' must be unique](https://github.com/mozilla-releng/balrog/blob/036d4a95110dfdf8f3e4bd812f5ac5c6351d3912/src/auslib/db.py#L1742), yet the scheduled rule change endpoints accept the idea of a new rule being scheduled with an alias that already exists. I highly doubt these rules would successfully be enacted, but regardless we should not accept their scheduling in the first place.

![image](https://github.com/mozilla-releng/balrog/assets/49649/dc54ef5c-2ba8-4587-88e5-492d506a3118)

**Comment by Stanzin7** [comment]
Hello @gabrielBusta @bhearsum, I am in the contribution phase for Outreachy. Can you please assign this task to me. Thanks )

**Comment by godplayer56** [comment]
Hey @bhearsum @gabrielBusta , I have some doubts regarding this issue -
If a new rule is being scheduled, no other aliases can schedule the same rule
Am I right about this?

**Comment by godplayer56** [comment]
Hey @bhearsum @gabrielBusta , I have some doubts regarding this issue -
If a new rule is being scheduled, no other aliases can schedule the same rule
Am I right about this?

**Comment by bhearsum** [comment]
> Hey @bhearsum @gabrielBusta , I have some doubts regarding this issue - If a new rule is being scheduled, no other aliases can schedule the same rule Am I right about this?

Yes - only one rule can exist with the same alias at any given time. This works correctly for the `rules` table and endpoints, which reject duplicate alias'. The scheduled rules table does _not_ have any constraints about unique alias' though (this is expected). We need some safeguards at the API level (https://github.com/mozilla-releng/balrog/blob/main/src/auslib/web/admin/views/rules.py#L188) to implement this.

**Comment by godplayer56** [comment]
Yes @bhearsum I think I got the fix. I am working on it

**Comment by bhearsum** [comment]
This is fixed by #3010. @michellemounde - I'm happy to help you work through your own fix for it as well, if you like, but if you'd prefer to focus on something else that's not yet fixed that's fine too.

**Comment by jcristau** [comment]
Reopening because #3010 is being reverted in #3048.

**Comment by godplayer56** [comment]
@jcristau Should I do some changes to my commit so as to fix this issue? or should I approach it differently?

**Comment by bhearsum** [comment]
Just to summarize: I think the tests we have for `/rules` and `/scheduled_changes/rules` are fine. However, I neglected to notice that we weren't testing `/rules/:id` and `/scheduled_changes/rules/:id` when I reviewed the patch.

Here's [a commit with a couple of tests as a starting point for this](https://github.com/mozilla-releng/balrog/commit/0b058e4044ff1532a37b097da73a53eb2f21de5b), which both currently fail for different reasons:
```
    def testPostRuleWithUnchangingAlias(self):
        """Verify that a rule with an alias can be updated, and that providing a value
        for alias that is the same as the current alias does not prevent this."""
    
        ret = self._post(
            "/rules/2",
            data=dict(
                alias="frodo",
                backgroundRate=50,
                data_version=1,
            ),
        )
        self.assertEqual(ret.status_code, 200, "Status Code: %d, Data: %s" % (ret.status_code, ret.get_data()))
        load = ret.get_json()
        self.assertEqual(load["new_data_version"], 2)
        r = dbo.rules.t.select().where(dbo.rules.rule_id == 1).execute().fetchall()
        self.assertEqual(len(r), 1)
>       self.assertEqual(r[0]["alias"], "frodo")
E       AssertionError: None != 'frodo'

tests/admin/views/test_rules.py:682: AssertionError
__ TestRuleScheduledChanges.testAddScheduledChangeExistingRuleUnchangingAlias __
[gw50] linux -- Python 3.9.18 /app/.tox/py39/bin/python

self = <tests.admin.views.test_rules.TestRuleScheduledChanges testMethod=testAddScheduledChangeExistingRuleUnchangingAlias>

    @mock.patch("time.time", mock.MagicMock(return_value=300))
    def testAddScheduledChangeExistingRuleUnchangingAlias(self):
        data = {
            "rule_id": 2,
            "telemetry_product": None,
            "telemetry_channel": None,
            "telemetry_uptake": None,
            "alias": "frodo",
            "backgroundRate": 50,
            "data_version": 1,
            "change_type": "update",
            "when": 1234567,
        }
        ret = self._post("/scheduled_changes/rules", data=data)
>       self.assertEqual(ret.status_code, 200, ret.get_data())
E       AssertionError: 400 != 200 : b'{"type": "about:blank", "title": "Bad Request", "detail": "Rule with alias exists", "status": 400}'
```

(The second is failing in the same way as reported in #3048, which is a good thing.)

We'll want some additional tests added here for the scheduled changes endpoint that utilizes other `change_type`s, as well as tests that update an existing scheduled change. I expect this would look similar to https://github.com/mozilla-releng/balrog/blob/cbc19776b17060e8d10160319e6016496fdbbe03/tests/admin/views/test_rules.py#L2286, but update `rule_id` `2` like the test I linked to.

**Comment by godplayer56** [comment]
Ok got it!  I am in a crunch for a couple of days 😅 . I will try to add a test for this as soon as possible 😄
